graphbolt/src/cuda/neighbor_sampler_impl.cu

```diff
@@ -0,0 +1,250 @@
+      curand_init(kCurandSeed, random_seed, indices[in_idx], &rng);
+    }
+
+    const auto rnd = curand_uniform(&rng);
+    const auto prob = weights ? weights[in_idx] : static_cast<weights_t>(1);
+    const auto exp_rnd = -__logf(rnd);
+    const float_t adjusted_prob =
+        prob > 0 ? exp_rnd / prob : std::numeric_limits<float_t>::infinity();
+    output[i] = {row_position, adjusted_prob};
+    edge_ids[i] = in_idx;
+
+    i += stride;
+  }
+}
+
+template <typename indices_t, typename float_t>
+struct decomposer_t {
+  __host__ __device__ ::cuda::std::tuple<indices_t&, float_t&> operator()(
+      ::cuda::std::tuple<indices_t, float_t>& key) const {
+    auto& [t, prob] = key;
+    return {t, prob};
+  }
+};
+
+template <typename indptr_t>
+struct MinInDegreeFanout {
+  const indptr_t* in_degree;
+  int64_t fanout;
+  __host__ __device__ auto operator()(int64_t i) {
+    return static_cast<indptr_t>(
+        min(static_cast<int64_t>(in_degree[i]), fanout));
+  }
+};
+
+template <typename indptr_t, typename indices_t>
+struct IteratorFunc {
+  indptr_t* indptr;
+  indices_t* indices;
+  __host__ __device__ auto operator()(int64_t i) { return indices + indptr[i]; }
+};
+
+template <typename indices_t>
+struct ConvertToBytes {
+  template <typename indptr_t>
+  __host__ __device__ auto operator()(indptr_t num_elements) {
+    return num_elements * sizeof(indices_t);
+  }
+};
+
+c10::intrusive_ptr<sampling::FusedSampledSubgraph> SampleNeighbors(
+    torch::Tensor indptr, torch::Tensor indices, torch::Tensor nodes,
+    const std::vector<int64_t>& fanouts, bool replace, bool layer,
+    bool return_eids, torch::optional<torch::Tensor> type_per_edge,
+    torch::optional<torch::Tensor> probs_or_mask,
+    torch::optional<int64_t> random_seed) {
+  TORCH_CHECK(
+      fanouts.size() == 1, "Heterogenous sampling is not supported yet!");
+  TORCH_CHECK(!replace, "Sampling with replacement is not supported yet!");
+  // We assume that indptr, indices, nodes, type_per_edge and probs_or_mask
+  // are all resident on the GPU. If not, it is better to first extract them.
+  const auto num_rows = nodes.size(0);
+  const auto fanout =
+      fanouts[0] >= 0 ? fanouts[0] : std::numeric_limits<int64_t>::max();
+  auto in_degree_and_sliced_indptr = SliceCSCIndptr(indptr, nodes);
+  auto in_degree = std::get<0>(in_degree_and_sliced_indptr);
+  auto sliced_indptr = std::get<1>(in_degree_and_sliced_indptr);
+  auto sub_indptr = ExclusiveCumSum(in_degree);
+  auto output_indptr = torch::empty_like(sub_indptr);
+  auto coo_rows = CSRToCOO(sub_indptr, indices.scalar_type());
+  const auto num_edges = coo_rows.size(0);
+  auto allocator = cuda::GetAllocator();
+  const auto stream = cuda::GetCurrentStream();
+  if (!random_seed.has_value()) {
+    random_seed = RandomEngine::ThreadLocal()->RandInt(
+        static_cast<int64_t>(0), std::numeric_limits<int64_t>::max());
+  }
+  torch::Tensor picked_eids;
+
+  AT_DISPATCH_INTEGRAL_TYPES(
```

Comment:
nit: can you polish the code a little bit according to the comments I made to the previous PR?
- Add blank line between section, comment each section
- Keep the comment neutral and objective, avoiding subjective pronouns like "we" and "you."
- Pick better name for the local variables, if the name is not contextual enough, add comment.